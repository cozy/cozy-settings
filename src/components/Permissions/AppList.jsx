import React from 'react'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import Icon from 'cozy-ui/transpiled/react/Icon'
import Typography from 'cozy-ui/transpiled/react/Typography'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon, {
  smallSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { routes } from 'constants/routes'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemSecondaryAction'
import NavigationList, {
  NavigationListSection
} from 'cozy-ui/transpiled/react/NavigationList'
import { isQueryLoading, hasQueryBeenLoaded } from 'cozy-client'
import { buildAppsQuery, buildKonnectorsQuery } from 'lib/queries'
import { useQuery } from 'cozy-client'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

const AppList = () => {
  const { t } = useI18n()
  const appsQuery = buildAppsQuery()
  const queryResultApps = useQuery(appsQuery.definition, appsQuery.options)
  const konnectorsQuery = buildKonnectorsQuery()
  const queryResultKonnectors = useQuery(
    konnectorsQuery.definition,
    konnectorsQuery.options
  )
  const { isMobile, isTablet } = useBreakpoints()
  const toNotDisplay = ['home', 'store', 'settings']

  return (
    <Page
      className={isMobile || isTablet ? '' : 'u-maw-7'}
      withoutMarginTop={isMobile || isTablet}
    >
      <PageTitle>{t('Permissions.applications')}</PageTitle>
      {(isQueryLoading(queryResultApps) ||
        isQueryLoading(queryResultKonnectors)) &&
      (!hasQueryBeenLoaded(queryResultApps) ||
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' ||
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <NavigationList>
          <NavigationListSection>
            {queryResultApps.data
              .concat(queryResultKonnectors.data)
              .sort((a, b) => a.slug.localeCompare(b.slug))
              .filter(a => !toNotDisplay.includes(a.slug))
              .map(appOrKonnector => {
                return (
                  <div key={appOrKonnector.name}>
                    <ListItem
                      button
                      component="a"
                      href={`/#${routes.appList}/${appOrKonnector.slug}`}
                    >
                      <ListItemIcon>
                        <AppIcon app={appOrKonnector} />
                      </ListItemIcon>
                      <ListItemText
                        primary={appOrKonnector.name}
                        secondary={t('Permissions.numberOfPermissions', {
                          smart_count: Object.entries(
                            appOrKonnector.permissions
                          ).length
                        })}
                      />
                      <ListItemSecondaryAction>
                        <Icon
                          icon={RightIcon}
                          size={smallSize}
                          className="u-mr-1"
                          style={{ color: 'var(--secondaryTextColor)' }}
                        />
                      </ListItemSecondaryAction>
                    </ListItem>
                    <Divider variant="inset" />
                  </div>
                )
              })}
          </NavigationListSection>
        </NavigationList>
      )}
    </Page>
  )
}

export default AppList
