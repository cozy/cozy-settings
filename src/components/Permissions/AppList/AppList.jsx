import React from 'react'

import Icon from 'cozy-ui/transpiled/react/Icon'
import Typography from 'cozy-ui/transpiled/react/Typography'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'
import ListItem from 'cozy-ui/transpiled/react/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon, { smallSize } from 'cozy-ui/transpiled/react/ListItemIcon'
import Divider from 'cozy-ui/transpiled/react/Divider'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/ListItemSecondaryAction'
import NavigationList, {
  NavigationListSection
} from 'cozy-ui/transpiled/react/NavigationList'

import { routes } from 'constants/routes'
import useAppsOrKonnectors from 'components/Permissions/hooks/useAppsOrKonnectors'

const AppList = () => {
  const { t } = useI18n()

  const { isResultLoading, hasQueryFailed, appsResult, konnectorsResult } =
    useAppsOrKonnectors()

  const toNotDisplay = ['home', 'store', 'settings']

  return (
    <>
      {isResultLoading ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : hasQueryFailed ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <NavigationList>
          <NavigationListSection>
            {appsResult.data
              .concat(konnectorsResult.data)
              .sort((a, b) => a.slug.localeCompare(b.slug))
              .filter(a => !toNotDisplay.includes(a.slug))
              .map(appOrKonnector => {
                return (
                  <div key={appOrKonnector.name}>
                    <ListItem
                      button
                      component="a"
                      href={`/#${routes.appList}/${appOrKonnector.slug}`}
                    >
                      <ListItemIcon>
                        <AppIcon app={appOrKonnector} />
                      </ListItemIcon>
                      <ListItemText
                        primary={appOrKonnector.name}
                        secondary={t('Permissions.numberOfPermissions', {
                          smart_count: Object.entries(
                            appOrKonnector.permissions
                          ).length
                        })}
                      />
                      <ListItemSecondaryAction>
                        <Icon
                          icon={RightIcon}
                          size={smallSize}
                          className="u-mr-1"
                          style={{ color: 'var(--secondaryTextColor)' }}
                        />
                      </ListItemSecondaryAction>
                    </ListItem>
                    <Divider variant="inset" />
                  </div>
                )
              })}
          </NavigationListSection>
        </NavigationList>
      )}
    </>
  )
}

export default AppList
