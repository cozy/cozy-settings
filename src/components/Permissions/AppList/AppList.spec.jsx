import { render } from '@testing-library/react'
import React from 'react'
import AppList from './AppList'
import { Q, useQuery, isQueryLoading, hasQueryBeenLoaded } from 'cozy-client'
import { TestI18n } from 'test/AppLike'
import { BreakpointsProvider } from 'cozy-ui/transpiled/react/providers/Breakpoints'

jest.mock('react-router-dom', () => {
  return {
    ...jest.requireActual('react-router-dom'),
    Link: ({ narrow, children }) => (
      <div data-testid="page" data-narrow={narrow}>
        {children}
      </div>
    )
  }
})

jest.mock('cozy-client')

jest.mock('components/Page', () => {
  // eslint-disable-next-line react/display-name
  return ({ narrow, children }) => (
    <div data-testid="page" data-narrow={narrow}>
      {children}
    </div>
  )
})

jest.mock('components/PageTitle', () => {
  // eslint-disable-next-line react/display-name
  return ({ children }) => <div data-testid="page-title">{children}</div>
})

jest.mock('cozy-ui/transpiled/react/Spinner', () => {
  // eslint-disable-next-line react/display-name
  return ({ size }) => <div data-testid="Spinner" data-size={size}></div>
})

jest.mock('cozy-ui/transpiled/react/NavigationList', () => {
  return {
    __esModule: true,
    default: ({ children }) => (
      <div data-testid="NavigationList">{children}</div>
    ),
    NavigationListSection: ({ children }) => (
      <div data-testid="NavigationListSection">{children}</div>
    )
  }
})

jest.mock('cozy-ui/transpiled/react/AppIcon', () => {
  // eslint-disable-next-line react/display-name
  return () => <div data-testid="AppIcon"></div>
})

jest.mock('cozy-ui/transpiled/react/ListItemText', () => {
  // eslint-disable-next-line react/display-name
  return ({ primary, secondary }) => (
    <div
      data-testid="ListItemText"
      data-primary={primary}
      data-secondary={secondary}
    ></div>
  )
})

describe('AppList', () => {
  beforeEach(() => {
    const queryResultApps = {
      fetchStatus: 'loaded',
      data: [
        {
          permissions: {
            files: {
              type: 'io.cozy.files',
              description: 'Required to access the files'
            },
            allFiles: {
              type: 'io.cozy.files.*',
              description: 'Required to access the files'
            },
            apps: {
              type: 'io.cozy.apps',
              description:
                'Required by the cozy-bar to display the icons of the apps',
              verbs: ['GET']
            }
          },
          slug: 'contacts',
          name: 'Contacts'
        },
        {
          permissions: {},
          slug: 'home',
          name: 'Home'
        },
        {
          permissions: {},
          slug: 'settings',
          name: 'Settings'
        },
        {
          permissions: {},
          slug: 'store',
          name: 'Store'
        }
      ]
    }
    const queryResultKonnectors = {
      fetchStatus: 'loaded',
      data: [
        {
          permissions: {
            apps: {
              type: 'io.cozy.apps',
              description:
                'Required by the cozy-bar to display the icons of the apps',
              verbs: ['GET']
            }
          },
          slug: 'alan',
          name: 'Alan'
        }
      ]
    }
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(true)
    Q.mockReturnValue({ getById: () => 'kfrf' })
    useQuery.mockReturnValueOnce(queryResultApps)
    useQuery.mockReturnValueOnce(queryResultKonnectors)
    useQuery.mockReturnValueOnce(queryResultApps)
    useQuery.mockReturnValueOnce(queryResultKonnectors)
  })
  it('should display appName when query has been loaded', () => {
    hasQueryBeenLoaded.mockReturnValue(true)
    const { container } = render(
      <BreakpointsProvider>
        <TestI18n>
          <AppList />
        </TestI18n>
      </BreakpointsProvider>
    )
    expect(container).toMatchSnapshot()
  })

  it('should render a spinner when query is loading and has not been loaded', () => {
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(false)
    const { queryByTestId } = render(
      <BreakpointsProvider>
        <TestI18n>
          <AppList />
        </TestI18n>
      </BreakpointsProvider>
    )
    expect(queryByTestId('Spinner')).toBeTruthy()
  })

  it('should display sorted app and konnector names when query is not loading', () => {
    isQueryLoading.mockReturnValue(false)
    const { getAllByTestId, container } = render(
      <BreakpointsProvider>
        <TestI18n>
          <AppList />
        </TestI18n>
      </BreakpointsProvider>
    )
    expect(container).toMatchSnapshot()
    expect(getAllByTestId('ListItemText')[1]).toHaveAttribute(
      'data-primary',
      'Contacts'
    )
    expect(getAllByTestId('ListItemText')[0]).toHaveAttribute(
      'data-primary',
      'Alan'
    )
  })

  it('should not display Home, Settings and Store when query is not loading', () => {
    isQueryLoading.mockReturnValue(false)
    const { queryByText } = render(
      <BreakpointsProvider>
        <TestI18n>
          <AppList />
        </TestI18n>
      </BreakpointsProvider>
    )
    expect(queryByText('Home')).toBeFalsy
    expect(queryByText('Settings')).toBeFalsy
    expect(queryByText('Store')).toBeFalsy
  })

  it('should render Permissions.failedRequest when query status is failed', () => {
    useQuery.mockReset()
    useQuery.mockReturnValue({ fetchStatus: 'failed' })
    const { queryByText } = render(
      <BreakpointsProvider>
        <TestI18n>
          <AppList />
        </TestI18n>
      </BreakpointsProvider>
    )
    expect(queryByText('The request has failed')).toBeTruthy()
  })
})
