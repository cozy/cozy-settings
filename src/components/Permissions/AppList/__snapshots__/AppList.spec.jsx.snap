// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppList should display appName when query has been loaded 1`] = `
<div>
  <div
    data-testid="page"
  >
    <div
      data-testid="NavigationList"
    >
      <div
        data-testid="NavigationListSection"
      >
        <div>
          <li
            class="MuiListItem-container"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction"
              href="/#/permissions/slug/alan"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root"
              >
                <div
                  data-testid="AppIcon"
                />
              </div>
              <div
                data-primary="Alan"
                data-secondary="1 permission"
                data-testid="ListItemText"
              />
              <span
                class="MuiTouchRipple-root"
              />
            </a>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <svg
                class="u-mr-1 styles__icon___23x3R"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </li>
          <hr
            class="MuiDivider-root MuiDivider-inset"
          />
        </div>
        <div>
          <li
            class="MuiListItem-container"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction"
              href="/#/permissions/slug/contacts"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root"
              >
                <div
                  data-testid="AppIcon"
                />
              </div>
              <div
                data-primary="Contacts"
                data-secondary="3 permissions"
                data-testid="ListItemText"
              />
              <span
                class="MuiTouchRipple-root"
              />
            </a>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <svg
                class="u-mr-1 styles__icon___23x3R"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </li>
          <hr
            class="MuiDivider-root MuiDivider-inset"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppList should display sorted app and konnector names when query is not loading 1`] = `
<div>
  <div
    data-testid="page"
  >
    <div
      data-testid="NavigationList"
    >
      <div
        data-testid="NavigationListSection"
      >
        <div>
          <li
            class="MuiListItem-container"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction"
              href="/#/permissions/slug/alan"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root"
              >
                <div
                  data-testid="AppIcon"
                />
              </div>
              <div
                data-primary="Alan"
                data-secondary="1 permission"
                data-testid="ListItemText"
              />
              <span
                class="MuiTouchRipple-root"
              />
            </a>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <svg
                class="u-mr-1 styles__icon___23x3R"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </li>
          <hr
            class="MuiDivider-root MuiDivider-inset"
          />
        </div>
        <div>
          <li
            class="MuiListItem-container"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction"
              href="/#/permissions/slug/contacts"
              tabindex="0"
            >
              <div
                class="MuiListItemIcon-root"
              >
                <div
                  data-testid="AppIcon"
                />
              </div>
              <div
                data-primary="Contacts"
                data-secondary="3 permissions"
                data-testid="ListItemText"
              />
              <span
                class="MuiTouchRipple-root"
              />
            </a>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <svg
                class="u-mr-1 styles__icon___23x3R"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M12.707 7.293l-6-6a.999.999 0 10-1.414 1.414L10.586 8l-5.293 5.293a.999.999 0 101.414 1.414l6-6a.999.999 0 000-1.414"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </li>
          <hr
            class="MuiDivider-root MuiDivider-inset"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
